taichi/transforms/make_thread_local.cpp

```diff
@@ -0,0 +1,153 @@
+#include "taichi/ir/ir.h"
+#include "taichi/ir/transforms.h"
+#include "taichi/ir/analysis.h"
+#include "taichi/ir/visitors.h"
+
+TLANG_NAMESPACE_BEGIN
+
+namespace {
+
+bool is_atomic_op_linear(AtomicOpType op_type) {
+  return op_type == AtomicOpType::add || op_type == AtomicOpType::sub;
+}
```

Comment:
Why only add/sub? I'm sure we would see huge speed ups applying this to max/min and others as well. I think we will need to load the original value to thread local pointer in the prologue instead of zero fill.

Reply:
Maybe just nit, but statement $ numbers are all out of order for the prologue and epilogue

```
kernel {
  $0 = offloaded range_for(0, 1048576) block_dim=adaptive  
  prologue {
    <i32 x1> $64 = thread local ptr (offset = 0 B)
    <i32 x1> $65 = const [0]
    <i32 x1> $66 : global store [$64 <- $65]
  }
  body {
    <i32 x1> $1 = thread local ptr (offset = 0 B)
    <i32 x1> $2 = loop $0 index 0
    <gen*x1> $3 = get root
    <i32 x1> $4 = const [0]
    <gen*x1> $5 = [S0root][root]::lookup($3, $4) activate = false
    <gen*x1> $6 = get child [S0root->S1dense] $5
    <gen*x1> $7 = [S1dense][dense]::lookup($6, $2) activate = false
    <i32*x1> $8 = get child [S1dense->S2place_i32] $7
    <i32 x1> $9 = global load $8
    <i32 x1> $10 = global load $1
    <i32 x1> $11 = add $10 $9
    <i32 x1> $12 : global store [$1 <- $11]
  }
  epilogue {
    <i32 x1> $68 = thread local ptr (offset = 0 B)
    <i32 x1> $69 = global load $68
    <gen*x1> $82 = get root
    <i32 x1> $98 = const [0]
    <gen*x1> $84 = [S0root][root]::lookup($82, $98) activate = false
    <gen*x1> $85 = get child [S0root->S3dense] $84
    <gen*x1> $87 = [S3dense][dense]::lookup($85, $98) activate = false
    <i32*x1> $88 = get child [S3dense->S4place_i32] $87
    <i32 x1> $71 = atomic add($88, $69)
  }
}
```

Reply:
> Why only add/sub? I'm sure we would see huge speed ups applying this to max/min and others as well.

Yeah we can add that later. I'm using add/sub just because it's easier to test.



> I think we will need to load the original value to thread local pointer in the prologue instead of zero fill.

Actually I think zero-fill leads to the correct behavior. We are only accumulating local contributions here, and local contributions start with 0.

Reply:
Right, I just meant you may need to load the original value to get min/max to work. Sounds good though thanks!

Reply:
> the $ numbers are all out of order for the prologue and epilogue

Maybe we should just use `BasicStmtVisitor` instead of `IRVisitor` for `re_id`...